File: go/path_encoding_test.go
// Regression coverage for the generator emitting url.PathEscape on every
// string :path param (not just the one endpoint that originally needed it):
// a raw slash, space, or non-ASCII character in a string id must reach the
// wire as one escaped path segment, not split into extra segments or sent
// unescaped. r.URL.Path (net/http's own decoded view) cannot tell the two
// cases apart - a literal '/' used as a real separator and an escaped
// "%2F" both decode back to the same string - so this asserts on
// r.RequestURI, the literal bytes that were sent on the wire.
package sdk_test

import (
	"context"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func TestStringPathParamsAreURLEncoded(t *testing.T) {
	const rawCopysetID = "copyset/id with spaces/café"
	const wantEscaped = "copyset%2Fid%20with%20spaces%2Fcaf%C3%A9"

	var requestURI string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		requestURI = r.RequestURI
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"status":"success","message":"ok","data":{"id":"` + rawCopysetID + `","storageId":"s1","state":"active"}}`))
	}))
	defer srv.Close()
	client := newTestClient(t, srv.URL)

	copyset, err := client.Storages.GetCopysetStatus(context.Background(), 7, rawCopysetID)
	if err != nil {
		t.Fatalf("GetCopysetStatus: %v", err)
	}
	if copyset.ID != rawCopysetID {
		t.Fatalf("got copyset id %q, want %q", copyset.ID, rawCopysetID)
	}
	if !strings.Contains(requestURI, wantEscaped) {
		t.Fatalf("request URI %q does not contain escaped segment %q", requestURI, wantEscaped)
	}
	if strings.Contains(requestURI, "/copysets/copyset/") {
		t.Fatalf("request URI %q sent the '/' unescaped, splitting the path", requestURI)
	}
}

File: ts/test/path_encoding.test.ts
// Regression coverage for the generator emitting encodeURIComponent on
// every string :path param (not just the one endpoint that originally
// needed it). Exercises the real fetch-based transport (createServerClient),
// not the fake RequestFn used elsewhere in this directory: fetch's own URL
// parser treats a raw '/' as a structural separator (splitting the path
// into extra segments) while auto-escaping other unsafe characters, so only
// the literal bytes actually sent on the wire (a plain http.Server's
// req.url, which Node never auto-decodes) can tell an escaped "%2F" apart
// from a real segment boundary.
import { test } from 'node:test'
import assert from 'node:assert/strict'
import http from 'node:http'
import type { AddressInfo } from 'node:net'
import { createServerClient } from '../dist/index.js'

test('string path params are URL-encoded on the wire', async () => {
  const rawCopysetId = 'copyset/id with spaces/café'

  let requestURL = ''
  const server = http.createServer((req, res) => {
    requestURL = req.url ?? ''
    res.setHeader('content-type', 'application/json')
    res.end(JSON.stringify({
      status: 'success', message: 'ok',
      data: { id: rawCopysetId, storageId: 's1', state: 'active' },
    }))
  })
  await new Promise<void>(resolve => server.listen(0, resolve))
  const { port } = server.address() as AddressInfo

  const privateKey = Buffer.from(new Uint8Array(32).fill(7)).toString('base64')
  const client = createServerClient({ baseUrl: `http://127.0.0.1:${port}`, privateKey })

  try {
    const copyset = await client.storages.getCopysetStatus(7, rawCopysetId)
    assert.equal(copyset.id, rawCopysetId)
    assert.match(requestURL, /copyset%2Fid%20with%20spaces%2Fcaf%C3%A9/)
    assert.doesNotMatch(requestURL, /\/copysets\/copyset\//)
  } finally {
    await new Promise<void>(resolve => server.close(() => resolve()))
  }
})
